File: ch8/forkdemo3.c
#include<stdio.h>
#include<stdlib.h>
#include<unistd.h>

int main(int argc, char const *argv[])
{
    int fork_rv;

    printf("Before my pid is %d\n",getpid());

    fork_rv = fork();

    if (fork_rv == -1)
    {
        perror("fork");
    }
    else if (fork_rv == 0)
    {
        printf("I am the chlid, my pid is %d\n",getpid());
    }
    else
    {
        printf("I am the parent,my chlid id is %d\n",fork_rv);
    }
    
    return 0;
}

File: ch8/ch8_src/src/bin/forkdemo3.rs
use std::error::Error;

use nix::unistd::{fork, getpid, ForkResult};

fn main()->Result<(),Box<dyn Error>>{
    println!("Before my pid is {}\n",getpid());

    let fork_rv = unsafe { fork() }?;
    
    match fork_rv {
        ForkResult::Parent {  child } =>{
            println!("I am the parent,my chlid id is {}",child);
        },
        ForkResult::Child =>{
            println!("I am the chlid, my pid is {}",getpid());
        }
    }
    Ok(())
}